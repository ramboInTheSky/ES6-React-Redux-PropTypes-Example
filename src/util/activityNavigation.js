import findIndex from 'ramda/src/findIndex';
import propEq from 'ramda/src/propEq';

export default (activities = [], currentId) => {
  const currentActivityIndex = findIndex(propEq('id', currentId))(activities);
  const isLast = () => (activities.length ? currentActivityIndex + 1 === activities.length : false);
  const isFirst = () => (activities.length ? currentActivityIndex === 0 : false);

  const getNextHistoryId = () => {
    const isLastItem = isLast();
    const nextHistoryActivity =
      currentActivityIndex >= 0 && !isLastItem ? activities[currentActivityIndex + 1] : null;
    return nextHistoryActivity ? nextHistoryActivity.id : null;
  };

  const getPreviousHistoryId = () => {
    const isFirstItem = isFirst();

    const previousHistoryActivity =
      currentActivityIndex >= 0 && !isFirstItem ? activities[currentActivityIndex - 1] : null;
    return previousHistoryActivity ? previousHistoryActivity.id : null;
  };

  return {
    getNextHistoryId,
    getPreviousHistoryId,
    isFirst,
    isLast,
  };
};
