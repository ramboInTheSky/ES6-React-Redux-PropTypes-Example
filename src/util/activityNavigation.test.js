import activityNavigation from './activityNavigation';

const activities = [{ id: 'first' }, { id: 'second' }];

describe('activityNavigation', () => {
  it('should return 4 methods', () => {
    expect(activityNavigation(activities, 'first')).toMatchSnapshot();
  });

  describe('isFirst', () => {
    it('should return true if the item is the first activity', () => {
      expect(activityNavigation(activities, 'first').isFirst()).toBe(true);
    });
    it('should return false if the item is not the first activity', () => {
      expect(activityNavigation(activities, 'second').isFirst()).toBe(false);
    });
  });

  describe('isLast', () => {
    it('should return true if the item is the last activity', () => {
      expect(activityNavigation(activities, 'second').isLast()).toBe(true);
    });
    it('should return false if the item is not the last activity', () => {
      expect(activityNavigation(activities, 'first').isLast()).toBe(false);
    });
  });

  describe('getNextHistoryId', () => {
    it('should return the next id if currentId is not the last item', () => {
      expect(activityNavigation(activities, 'first').getNextHistoryId()).toEqual('second');
    });
    it('should return null if currentId is the last item', () => {
      expect(activityNavigation(activities, 'second').getNextHistoryId()).toBe(null);
    });
  });

  describe('getPreviousHistoryId', () => {
    it('should return the previous id if currentId is not the first item', () => {
      expect(activityNavigation(activities, 'second').getPreviousHistoryId()).toEqual('first');
    });
    it('should return null if currentId is the first item', () => {
      expect(activityNavigation(activities, 'first').getPreviousHistoryId()).toBe(null);
    });
  });
});
