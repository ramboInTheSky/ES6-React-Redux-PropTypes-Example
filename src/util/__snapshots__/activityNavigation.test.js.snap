// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activityNavigation should return 4 methods 1`] = `
Object {
  "getNextHistoryId": [Function],
  "getPreviousHistoryId": [Function],
  "isFirst": [Function],
  "isLast": [Function],
}
`;
